rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+*   **TF will become a suite of multiple packages, built by multiple authorities. What if the bugs get blamed on TF team**: With the modular model, we expect testing of 3rd party code to become easier. This can also be mitigated if the error messages are better, and if they can clearly point out which module the issue stems from. Finally, we can create an apple-swift like testing model, where we run a Jenkins setup that people can donate their machines to, and we can run continuous integration tests on their plugins.
+*   **Why not have APIs but still have a monolithic repository: **When everything is in a single repository, this enables developers to bypass the APIs, and depend on internals. Moreover, we cannot grant full control over different folders on our repository to our partners in a single repository. As long as they are in a single repository, they are still constrained by our build system and license. Finally, in a single repository we do not provide the option of closed source plugins for contributors.
+*   **Why not go with the OSS federation solutions?** OSS federation requires all dependencies to be in the federation before adding a repository. This is simply not possible for tensorflow, as eigen, llvm and many other dependencies will never be a part of the federation.
+*   **Documentation, how/where do we document everything?** With multiple repositories, structure of the documentation will need to be rethought, based on what is a part of "TensorFlow proper" and what is an optional feature.
+
+
+## Testing Plan
+
+We propose the following principles to be followed for testing in a modular world:
+
+* Each plugin tested separately.
+* Modules can plan their own integration tests.
+* Cut combinatorial explosion by divide and conquer.
+* Fuzzing at the core-module interface level if possible, in the case we need data marshalling between layers.
+* With this proposal, we aim to also simplify testing of tensorflow code. The biggest gain we expect in the modular world will be, we will be able to "divide and conquer" the TensorFlow support matrix.
+* Following describes an early modularized tensorflow package structure:
+
+
+![alt_text](20190305-modular-tensorflow/initial_tf_deps.png "Initial package structure")
+
+In the current setup, we need to test all of the above packages for different Python versions, operating systems, accelerators (CPU, GPU), compilers, and more variants combined. In the modularized world, each of these packages only need to be unit tested for the following:
+
+
+*   tensorflow-base: Operating systems, compiler versions and python versions only with CPU 
+*   tf-gpu: With GPU only, for different operating systems.
+*   tf-estimator: Only for different python versions
+
+When testing a package that has dependencies, such as tf-estimator, or tf-gpu, tensorflow-base will be installed with its latest stable release, to ensure to avoid any flakes by this package.
```

Comment:
How will CI be done? When someone submits a PR to TF Core, can plugins trigger CI to ensure that the PR doesn't cause regressions in the plugin?

Reply:
Each plugin shall manage their depending TF version by itself.

Reply:
CC @lamberta
It is actually a mix. There will have to be some plugins, like estimator we will declare as a part of core TF APIs, and document as such. For some other plugins, what @byronyi said will apply.

Reply:
No, our goal is for each repository to run their own unit tests only for PRs. Currently, TF's situation is becoming unmanageable, and it is a big developer concern on TF side.
Plugins will rely on TF core to keep its API syntactically and semantically compatible. Then they will run their own tests.
If the integration tests after PRs are merged uncover backwards incompatible changes in TF, or the other dependencies in the ecosystem these will be rolled back.